docs/utilities/recipes/ingredients.md

```diff
@@ -0,0 +1,173 @@
+  {
+    // Ingredient
+  },
+  {
+    // Custom ingredient
+    "type": "examplemod:example_ingredient"
+  }
+]
+```
+
+#### NBTIngredient
+
+`NBTIngredient`s compare the item, damage, and the share tags (as defined by `IForgeItem#getShareTag`) on an `ItemStack` for equivalency. This can be used by specifying the `type` as `forge:nbt`.
+
+```json5
+// For some input
+{
+  "type": "forge:nbt",
+  "item": "examplemod:example_item",
+  "nbt": {
+    // Add nbt data (must match exactly what is on the stack)
+  }
+}
+```
+
+### PartialNBTIngredient
+
+`PartialNBTIngredient`s are a looser version of [`NBTIngredient`][nbt] as they compare against a single or set of items and only keys specified within the tag. This can be used by specifying the `type` as `forge:partial_nbt`.
+
+```json5
+// For some input
+{
+  "type": "forge:partial_nbt",
+
+  // Either 'item' or 'items' must be specified
+  // If both are specified, only 'item' will be read
+  "item": "examplemod:example_item",
+  "items": [
+    "examplemod:example_item",
+    "examplemod:example_item2"
+    // ...
+  ],
+
+  "nbt": {
+    // Checks only for equivalency on 'key1' and 'key2'
+    // All other keys in the stack will not be checked
+    "key1": "data1",
+    "key2": {
+      // Data 2
+    }
+  }
+}
+```
+
+### IntersectionIngredient
+
+`IntersectionIngredient`s work as a set AND where the passed in stack must be within all supplied ingredients. There must be at least two ingredients supplied to this. This can be used by specifying the `type` as `forge:intersection`.
+
+```json5
+// For some input
+{
+  "type": "forge:intersection",
+
+  // All of these ingredients must return true to succeed
+  "children": [
+    {
+      // Ingredient 1
+    },
+    {
+      // Ingredient 2
+    }
+    // ...
+  ]
+}
+```
+
+### DifferenceIngredient
+
+`DifferenceIngredient`s work as a set subtraction (SUB) where the passed in stack must be within the first ingredient but not in the second ingredient. This can be used by specifying the `type` as `forge:difference`.
+
+```json5
+// For some input
+{
+  "type": "forge:difference",
+  "base": {
+    // Ingredient the stack is in
+  },
+  "subtracted": {
+    // Ingredient the stack is NOT in
+  }
+}
+```
+
+Creating Custom Ingredients
+---------------------------
+
+Custom ingredients can be created by implementing `IIngredientSerializer` for the created `Ingredient` subclass.
+
+!!! tip
+    Custom ingredients should subclass `AbstractIngredient` as it provides some useful abstractions for ease of implementation.
+
+### Ingredient Subclass
+
+There are three important methods to implement for each ingredient subclass:
+
+Method | Description
+:---: | :---
+getSerializer | Returns the [serializer] used to read and write the ingredient.
+test | Returns true if the input is valid for this ingredient.
+isSimple | Returns false if the data needed from the stack is mutable (such as tag or capability data). This should not be overridden if returning true.
```

Comment:
I don't think that is accurate. My understanding is `isSimple` returns `false` if the ingredient matchs on based on NBT, and `true` if you can match entirely based on a set of IDs. As such, an ingredient matching on tags is considered simple, but an NBT ingredient is not.

Also worth nothing `AbstractIngredient` currently abstracts `isSimple` as I assumed all custom ingredients need to consider whether to return `true` or `false`, might be worth making that not abstract if you think there is a case for keeping the default value (defaulting an abstract method shouldn't be a breaking change).